File: Golang/18.四数之和.go
package main

import (
	"fmt"
	"sort"
)
/*=======================================
* @FileName: 18.四数之和.go
* @Description: 计算和 target 相等的不重复四元数组
* @Author: TonyLaw
* @Date: 2021-08-23 01:25:41 Monday
* @Copyright:  © 2021 TonyLaw. All rights reserved.
=========================================*/
func fourSum(nums []int, target int) (quadruplets [][]int) {
	sort.Ints(nums)
	n := len(nums)
	for i := 0; i < n - 3 && nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] <= target; i++ {
		if i > 0 && nums[i] == nums[i - 1] || nums[i] + nums[n - 1] + nums[n - 2] + nums[n - 3] < target {
			continue
		}
		for j := i + 1; j < n - 2 && nums[i] + nums[j] + nums[j + 1] + nums[j + 2] <= target; j++ {
			if j > i + 1 && nums[j] == nums[j - 1] || nums[i] + nums[j] + nums[n - 2] + nums[n - 1] < target {
				continue  // 因为左边两个是最小值（可以向右移动增大）， 右边两位是最大值（固定了，移动不了）， 四数之和比 target 小，说明左边两位可以继续加大
			}
			for left, right := j + 1, n - 1; left < right; {
				if sum := nums[i] + nums[j] + nums[left] + nums[right]; sum == target {
					quadruplets = append(quadruplets, []int{nums[i], nums[j], nums[left], nums[right]})
					for left++; left < right && nums[left] == nums[left - 1]; left++ {
					}
					for right--; left < right && nums[right] == nums[right + 1]; right-- {
					}
				} else if sum < target {
					left++
				} else {
					right--
				}
			}
		}
	} 
	return
}


func main() {
	var i1 = []int{1, 0, -1, 0, -2, 2}
	var i2 = 0
	fourSum(i1, i2)
	fmt.Println(fourSum(i1, i2))
}

File: rust/18四数之和.rs
/*=======================================
* @FileName: 18四数之和.rs
* @Description: 
* @Author: TonyLaw
* @Date: 2021-09-01 23:10:39 Wednesday
* @Copyright:  © 2021 TonyLaw. All Rights Reserved.
=========================================*/
/*=======================================
                (题目难度：中等)
给你一个由 n 个整数组成的数组 nums ，和一个目标值 target 。请你找出并返回满足下述全部条件且不重复的四元组 [nums[a], nums[b], nums[c], nums[d]] ：

0 <= a, b, c, d < n
a、b、c 和 d 互不相同
nums[a] + nums[b] + nums[c] + nums[d] == target
你可以按 任意顺序 返回答案 。

 

示例 1：

输入：nums = [1,0,-1,0,-2,2], target = 0
输出：[[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]]
示例 2：

输入：nums = [2,2,2,2,2], target = 8
输出：[[2,2,2,2]]
 

提示：

1 <= nums.length <= 200
-109 <= nums[i] <= 109
-109 <= target <= 109
=========================================*/
struct Solution;

impl Solution {
    pub fn four_sum(mut nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let len = nums.len();
        if len < 4 {return vec![]};
        let mut ret = vec![];
        nums.sort_unstable();
        for i in 0..len - 3 {
            if nums[i] + nums[i + 1] + nums[i + 2] + nums[i + 3] > target { break; } //剪枝
            if nums[i] + nums[len - 3] + nums[len - 2] + nums[len - 1] < target { continue; } // 剪枝
            
            if i > 0 && nums[i] == nums[i - 1] { continue; } // 剪枝
            for j in i+1..len-2 {
                if nums[i] + nums[j] + nums[j + 1] + nums[j + 2] > target { break; } // 剪枝
                if nums[i] + nums[j] + nums[len - 2] + nums[len - 1] < target { continue; } // 剪枝

                if j > i + 1 && nums[j] == nums[j - 1] {continue; }  // 剪枝
                let (mut k, mut q) = (j + 1, len - 1);                
                loop {
                    if k >= q { break; }
                    if nums[i] + nums[j] + nums[k] + nums[k + 1] > target { break; } // 剪枝
                    if nums[i] + nums[j] + nums[q - 1] + nums[q] < target { break; } // 剪枝
                    if k > j + 1 && nums[k] == nums[k - 1] { k += 1; continue; } // 剪枝
                    if q > len - 1 && nums[q] == nums[k + 1] { q -= 1; continue; } // 剪枝
                    let sum = nums[i] + nums[j] + nums[k] + nums[q];
                    if sum == target {
                        ret.push(vec![nums[i], nums[j], nums[k], nums[q]]);
                        k += 1;
                    } else if sum > target { 
                        q -= 1;
                    } else {
                        k += 1;
                    }
                }
            }
        }                 
        return ret
    }
}

fn main() {
    let l1 = vec![1, 0, -1, 0, -2, 2];
    let n1 = 0;
    let result = Solution::four_sum(l1, n1);
    println!("{:?}", result);
}